client/trino-client/src/main/java/io/trino/client/spooling/SegmentsIterator.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client.spooling;
+
+import com.google.common.collect.AbstractIterator;
+import io.trino.client.CloseableIterator;
+import io.trino.client.QueryDataDecoder;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class SegmentsIterator
+        extends AbstractIterator<List<Object>>
+        implements CloseableIterator<List<Object>>
+{
+    private final SegmentLoader loader;
+    private final QueryDataDecoder decoder;
+    private final Deque<Segment> remainingSegments;
+    private CloseableIterator<List<Object>> currentIterator;
+
+    public SegmentsIterator(SegmentLoader loader, QueryDataDecoder decoder, List<Segment> remainingSegments)
+    {
+        this.loader = requireNonNull(loader, "loader is null");
+        this.decoder = requireNonNull(decoder, "decoder is null");
+        this.remainingSegments = new ArrayDeque<>(remainingSegments);
+    }
+
+    @Override
+    protected List<Object> computeNext()
+    {
+        if (currentIterator == null) {
+            if (!remainingSegments.isEmpty()) {
+                Segment segment = remainingSegments.removeFirst();
+                currentIterator = iterate(segment);
+            }
+            else {
+                return endOfData();
+            }
+        }
+
+        if (currentIterator.hasNext()) {
+            return currentIterator.next();
+        }
+
+        currentIterator = null;
+        return computeNext();
+    }
+
+    private CloseableIterator<List<Object>> iterate(Segment segment)
+    {
+        if (segment instanceof InlineSegment) {
+            return new InlineSegmentIterator((InlineSegment) segment, decoder);
+        }
+
+        if (segment instanceof SpooledSegment) {
+            return new SpooledSegmentIterator((SpooledSegment) segment, loader, decoder);
+        }
+
+        throw new UnsupportedOperationException("Unsupported segment type: " + segment.getClass().getName());
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        IOException exception = new IOException("Could not close all segments");
+        if (currentIterator != null) {
+            try {
+                currentIterator.close();
```

Comment:
This looks awkward. Why do you need to crate iterator to close it

Reply:
This drops the segments when the client is no longer interested in fetching the results (explicit close) or when the client receives an error while processing query (implicit close).

Reply:
Actually I meant to put this comment in 
```
iterate(segment).close();
```
line below.

The awkaward part is that to clean segments we need to get iterator for it and immediatelly close it. It would be more starightforward if you close segment directly.